fix(messages)!: decouple ext_messages from message grid

luukvbaal opened the pull request:
Problem:  ext_messages is implemented to mimic the message grid
          implementation w.r.t. scrolling messages, clearing scrolled
          messages, hit-enter-prompts and replacing a previous message.
          Meanwhile, an ext_messages UI may not be implemented in a way
          where these events are wanted. Moreover, correctness of these
          events even assuming a "scrolled message" implementation
          depends on fragile "currently visible messages" global state,
          which already isn't correct after a previous message was
          supposed to have been overwritten (because that should not only
          happen when `msg_scroll == false`).

Solution: 
- No longer attempt to keep track of the currently visible
            messages: remove the `msg_ext(_history)_visible` variables.
            UIs may remove messages pre-emptively (timer based), or never
            show messages that don't fit a certain area in the first place.
- No longer emit the `msg(_history)_clear` events to clear
            "scrolled" messages. This opens up the `msg_clear` event to
            be emitted when messages should actually be cleared (e.g.
            when the screen is cleared). May also be useful to emit before
            the first message in an event loop cycle as a hint to the UI
            that it is a new batch of messages (vim._extui currently
            schedules an event to determine that).
- Set `replace_last` explicitly at the few callsites that want
            this to be set to true to replace an incomplete status message.
- Don't store a "keep" message to be re-emitted.

Resolve #20416

luukvbaal commented:
Low effort PR mainly intended to discuss the posed problem-solution statement.

justinmk commented:
Looks nice!

> Not emitting the hit-return message also seems fair game. The handler can decide and trigger a hit-enter prompt itself (e.g. based on the message kind or text height of the message).

Also sounds nice.

What are the downsides?

luukvbaal commented:
> What are the downsides?

From my POV there aren't any, besides changing behavior which clients may be relying on (not sure who else to ping to confirm that). I would say a client is in a better position to decide whether clearing/prompting is required than to take these suggestions from Nvim.

justinmk commented:
we can merge it early in 0.11 development cycle and find out

JeanMertz commented:
> we can merge it early in 0.11 development cycle and find out

Given that the 0.11 dev cycle is underway, should this be merged?

clason commented:
Needs a rebase. What is the plan here (and for message grid rework generally)? Are we confident about this being ready for 0.11, or do we want to bump to 0.12?

folke commented:
I'm personally a big fan of this PR.

Will reduce `ui_attach` events a lot. In Noice I currently need to do all sorts of work-arounds and duplicate event detection to make sure the ui doesn't get spammed.

luukvbaal commented:
> duplicate event detection to make sure the ui doesn't get spammed.

I still haven't actually found a clear reproducer where `keep_msg` is responsible for this, do you have one? As mentioned in https://github.com/neovim/neovim/issues/20416#issuecomment-2484076484, we may still want to remove `keep_msg`. But I think the scope of this PR is too big, and I've since found that `replace_last` actually is useful sometimes (I think noice also uses it?).

Can still go through with at least disabling `keep_msg` and the `hit-enter-prompt` under the given reasoning. I'm no longer sure the breaking event changes here are worth it.

luukvbaal commented:
> I just did some testing and it actually seems most of those issues with re-submitting events seems to be solved already.

Yeah recursive/redundant events were already reduced in various other PRs. The `keep_msg` itself is quite rare(hence my struggle to reproduce it😅).

> However, as soon as I have the cmdline open, I get spammed with cmdline_show events.

Does this patch improve that?
```diff
diff --git a/src/nvim/ex_getln.c b/src/nvim/ex_getln.c
index 1eecee2a38..bc99a18086 100644
--- a/src/nvim/ex_getln.c
+++ b/src/nvim/ex_getln.c
@@ -3527,12 +3527,13 @@ void cmdline_ui_flush(void)
   CmdlineInfo *line = &ccline;
   while (level > 0 && line) {
     if (line->level == level) {
-      if (line->redraw_state == kCmdRedrawAll) {
+      CmdRedraw redraw_state = line->redraw_state;
+      line->redraw_state = kCmdRedrawNone;
+      if (redraw_state == kCmdRedrawAll) {
         ui_ext_cmdline_show(line);
-      } else if (line->redraw_state == kCmdRedrawPos) {
+      } else if (redraw_state == kCmdRedrawPos) {
         ui_call_cmdline_pos(line->cmdpos, line->level);
       }
-      line->redraw_state = kCmdRedrawNone;
       level--;
     }
     line = line->prev_ccline;
```

folke commented:
> Yeah recursive/redundant events were already reduced in various other PRs.

Should've checked that indeed :)

> Does this patch improve that?

Nope, still the same. `cmdline_show` events.

luukvbaal commented:
Depending on how you redraw, #32237 might help @folke. This completely eliminates redundant cmdline events for me, just a complete redraw (`<C-L>, :mode) still re-emits the cmdline state.

folke commented:
@luukvbaal yes, that fixes it indeed!

luukvbaal committed:
fix(messages): decouple ext_messages from message grid

Problem:  ext_messages is implemented to mimic the message grid
          implementation w.r.t. scrolling messages, clearing scrolled
          messages, hit-enter-prompts and replacing a previous message.
          Meanwhile, an ext_messages UI may not be implemented in a way
          where these events are wanted. Moreover, correctness of these
          events even assuming a "scrolled message" implementation
          depends on fragile "currently visible messages" global state,
          which already isn't correct after a previous message was
          supposed to have been overwritten (because that should not only
          happen when `msg_scroll == false`).
Solution: - No longer attempt to keep track of the currently visible
            messages: remove the `msg_ext(_history)_visible` variables.
            UIs may remove messages pre-emptively (timer based), or never
            show messages that don't fit a certain area in the first place.
          - No longer emit the `msg(_history)_clear` events to clear
            "scrolled" messages. This opens up the `msg_clear` event to
            be emitted when messages should actually be cleared (e.g.
            when the screen is cleared). May also be useful to emit before
            the first message in an event loop cycle as a hint to the UI
            that it is a new batch of messages (vim._extui currently
            schedules an event to determine that).
          - Set `replace_last` explicitly at the few callsites that want
            this to be set to true to replace an incomplete status message.
          - Don't store a "keep" message to be re-emitted.

luukvbaal commented:
After looking into this again to resolve #34035, I found that "currently visible messages", which is used to determine when to emit these events, can already get out-of-sync with what it would be on the message grid. The fact that it doesn't work properly in it's current form increased my confidence in making a breaking change here (are there any UIs that tried to implement this exactly as the message grid or do they all just implement `msg_set_pos` events?), so I've rebased and reworked this.

justinmk reviewed (approved):
Let's try it! 🚀  The simplification is worth a minor breakage.

Rebasing should fix the broken windows CI.

justinmk commented on runtime/doc/news.txt:
```diff
@@ -67,7 +67,8 @@ EDITOR
 
-• todo
+• |ui-messages| no longer emits the `msg_show.return_prompt`, `msg_clear` and
+  `msg_history_clear` events.
```
```suggestion
  `msg_history_clear` events. Performance benefit: reduces UI events traffic.
```

luukvbaal commented on runtime/doc/news.txt:
```diff
@@ -67,7 +67,10 @@ EDITOR
 
-• todo
+• |ui-messages| no longer emits the `msg_show.return_prompt`, `msg_clear` and
+  `msg_history_clear` events. These events arbitrarily assume a message UI
```
The guidance is for UIs to determine themselves when to clear messages🫣 @xiyaowong already seems to have the right idea.

If a UI opens a new window to show messages, it makes sense to just clear messages after a few seconds. If a UI shows messages in a persistent area (cmdline), it makes sense to "clear" messages at the start of a new batch of messages. extui uses the next eventloop cycle to determine this (vim.schedule), like the message grid implementation does.

kaixuan188 commented:
Using neovim-vscode, I wasted a ton of tokens because of this bug. I finally got Claude to help me find this PR... crash...
